### PR TITLE
Add @upgrade service

### PR DESCRIPTION
**Pull request by ericof**
- [X] Implement `@upgrade` endpoint
- [x] Document usage of `@upgrade`

### CODE REVIEW

**Comment by ericof** [comment]
@jenkins-plone-org please run jobs

File: src/plone/restapi/services/upgrade/post.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+        view = Upgrade(self.context, self.request)
+        versions = view.versions()
+        return {
+            "@id": f"{self.context.absolute_url()}/@upgrade",
+            "report": report,
+            "versions": {
+                "instance": versions["instance"],
+                "fs": versions["fs"],
+            },
+        }
```
Comment (davisagli) [inline src/plone/restapi/services/upgrade/post.py:28]: @ericof Should we return a clear status for whether the upgrade was successful? https://github.com/plone/Products.CMFPlone/blob/master/Products/CMFPlone/browser/templates/plone-upgrade.pt appears to check this based on whether the versions ended up equal or not.

File: src/plone/restapi/tests/test_services_upgrade.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+from Products.CMFCore.utils import getToolByName
+from Products.CMFPlone.factory import _DEFAULT_PROFILE
+
+import transaction
+import unittest
+
+
+# Python 3 is only supported on 5.2+.
+# This means you can not upgrade from 5.1 or earlier.
+START_PROFILE = "5200"
```
Comment (davisagli) [inline src/plone/restapi/tests/test_services_upgrade.py:14]: I'd call it START_VERSION

**Review by davisagli — COMMENTED** [review]
Looks good so far. Apart from docs, I have one suggestion to provide a `status` in the result, which would also make the test assertions a bit simpler.

File: src/plone/restapi/tests/test_services_upgrade.py (hunk 2 above)
Comment (ericof) [reply src/plone/restapi/tests/test_services_upgrade.py:14]: This was copied verbatim from MigrationTool tests. (But I agree with you)

**Comment by ericof** [comment]
@jenkins-plone-org please run jobs

File: docs/source/endpoints/upgrade.md [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+(upgrade)=
+
+# Upgrade
+
+A Plone site needs to be in sync with the version available on the file system.
+The `@upgrade` endpoint exposes upgrade information about the Plone backend, and supports running the upgrade of the site.
+
+```{note}
+The upgrade endpoint is protected by the `cmf.ManagePortal` permission that requires the Site Administrator or Manager role.
```
Comment (davisagli) [inline docs/source/endpoints/upgrade.md:18]: I don't think the Site Administrator role has "Manage portal". That's one of the key differences between Site Administrator and Manager.

File: docs/source/endpoints/upgrade.md [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+## Get upgrade information
+
+Send a `GET` request to the `@upgrade` endpoint:
+
+```{eval-rst}
+..  http:example:: curl httpie python-requests
+    :request: ../../../src/plone/restapi/tests/http-examples/upgrade_get.req
+```
+
+The response will contain the upgrade information:
```
Comment (davisagli) [inline docs/source/endpoints/upgrade.md:30]:
```suggestion
The response will contain information about available upgrade steps:
```

File: docs/source/endpoints/upgrade.md [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+## Dry run the upgrade
+
+Send a `POST` request to the `@upgrade` endpoint passing the value of `dry_run` as `true`:
+
+```{eval-rst}
+..  http:example:: curl httpie python-requests
+    :request: ../../../src/plone/restapi/tests/http-examples/upgrade_post_dry_run.req
+```
+
```
Comment (davisagli) [inline docs/source/endpoints/upgrade.md:79]:
```suggestion
A dry run runs the entire upgrade, but does not commit the result to the database.
```

File: docs/source/endpoints/upgrade.md [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+      }
+    ]
+  },
+  "versions": {
+    "fs": "6008",
+    "instance": "6006"
+  }
+}
+```
+
```
Comment (davisagli) [inline docs/source/endpoints/upgrade.md:70]:
```suggestion

`fs` shows the current version of the `CMFPlone:default` profile on the filesystem.
`instance` shows the current version of the `CMFPlone:default` profile in the database.

```

**Review by davisagli — CHANGES_REQUESTED** [review]

File: docs/source/endpoints/upgrade.md (hunk 3 above)
Comment (ericof) [reply docs/source/endpoints/upgrade.md]: Fixed

File: docs/source/endpoints/upgrade.md [hunk 7]
```diff
@@ -0,0 +1,125 @@
+---
+myst:
+  html_meta:
+    "description": "The @upgrade endpoint exposes  Plone backend."
```
Comment (stevepiercy) [inline docs/source/endpoints/upgrade.md:4]:
```suggestion
    "description": "The @upgrade endpoint exposes upgrade information about the Plone backend, and supports running the upgrade of the site."
```

File: docs/source/endpoints/upgrade.md [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+    "property=og:description": "The @system endpoint exposes system information about the Plone backend."
```
Comment (stevepiercy) [inline docs/source/endpoints/upgrade.md]:
```suggestion
    "property=og:description": "The @upgrade endpoint exposes upgrade information about the Plone backend, and supports running the upgrade of the site."
```

**Review by davisagli — APPROVED** [review]

File: docs/source/endpoints/upgrade.md [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+        "title": "Update plonetheme.barceloneta registry"
+      }
+    ]
+  },
+  "versions": {
+    "fs": "6008",
+    "instance": "6006"
+  }
+}
+```
```
Comment (stevepiercy) [inline docs/source/endpoints/upgrade.md:69]: I thought that the responses are supposed to be in separate files for testing, and referenced like this:

````
```{literalinclude} ../../../src/plone/restapi/tests/http-examples/upgrade_get.resp
:language: http
```
````

File: docs/source/endpoints/upgrade.md [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+
+## Dry run the upgrade
+
+Send a `POST` request to the `@upgrade` endpoint passing the value of `dry_run` as `true`:
+
+```{eval-rst}
+..  http:example:: curl httpie python-requests
+    :request: ../../../src/plone/restapi/tests/http-examples/upgrade_post_dry_run.req
+```
+A dry run runs the entire upgrade, but does not commit the result to the database.
```
Comment (stevepiercy) [inline docs/source/endpoints/upgrade.md:82]: Insert a blank line for code legibility.

```suggestion

A dry run runs the entire upgrade, but does not commit the result to the database.
```

File: docs/source/endpoints/upgrade.md [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+The response will contain the result of the upgrade operation, indicating `dry_run` was selected:
+
+```http
```
Comment (stevepiercy) [inline docs/source/endpoints/upgrade.md:85]: Same as above: put this in a separate file for testing and reference it.

File: docs/source/endpoints/upgrade.md [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+
+## Run upgrade
+
+Send a `POST` request to the `@upgrade` endpoint, with `dry_run` set to `false`:
+
+```{eval-rst}
+..  http:example:: curl httpie python-requests
+    :request: ../../../src/plone/restapi/tests/http-examples/upgrade_post.req
+```
+The response will contain the result of the upgrade operation:
```
Comment (stevepiercy) [inline docs/source/endpoints/upgrade.md:109]: Add a blank line.

```suggestion

The response will contain the result of the upgrade operation:
```

File: docs/source/endpoints/upgrade.md [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+
+```http
```
Comment (stevepiercy) [inline docs/source/endpoints/upgrade.md:111]: Move to a separate file.

File: news/1525.feature [hunk 14]
```diff
@@ -0,0 +1 @@
+Add @upgrade endpoint to use upgrade a Plone instance
```
Comment (stevepiercy) [inline news/1525.feature:1]:
```suggestion
Add @upgrade endpoint to preview or run an upgrade of a Plone instance
```

**Review by stevepiercy — CHANGES_REQUESTED** [review]
Minor tweaks, but I'm not clear whether the responses should be in the tests folder, like the other endpoints.

**Review by jensens — APPROVED** [review]

**Review by sneridagh — COMMENTED** [review]
Late to the party, but LGTM too!

**Pull request merged**